core/include/detray/utils/root_finding.hpp

```diff
@@ -0,0 +1,302 @@
+    scalar_t f_u{f(upper)};
+    std::size_t n_tries{0u};
+
+    // If there is no sign change in interval, we don't know if there is a root
+    while (!math::signbit(f_l * f_u)) {
+        // No interval could be found to bracket the root
+        // Might be correct, if there is not root
+        if ((n_tries == 1000u) || !std::isfinite(f_l) || !std::isfinite(f_u)) {
+#ifdef DEBUG
+            std::cout << "WARNING: Could not bracket a root" << std::endl;
+#endif
+            bracket = {a, b};
+            return false;
+        }
+        scalar_t d{k * (upper - lower)};
+        // Make interval larger in the direction where the function is smaller
+        if (math::fabs(f_l) < math::fabs(f_u)) {
+            lower -= d;
+            f_l = f(lower);
+        } else {
+            upper += d;
+            f_u = f(upper);
+        }
+        ++n_tries;
+    }
+
+    bracket = {lower, upper};
+    return true;
+}
+
+/// @brief Find a root using the Newton-Raphson and Bisection algorithms
+///
+/// @param s initial guess for the root
+/// @param evaluate_func evaluate the function and its derivative
+/// @param max_path don't consider root if it is too far away
+///
+/// @see Numerical Recepies pp. 445
+///
+/// @return pathlength to root and the last step size
+template <typename scalar_t, typename function_t>
+DETRAY_HOST_DEVICE inline std::pair<scalar_t, scalar_t> newton_raphson_safe(
+    function_t &evaluate_func, scalar_t s,
+    const scalar_t convergence_tolerance = 1.f * unit<scalar_t>::um,
+    const std::size_t max_n_tries = 1000u,
+    const scalar_t max_path = 5.f * unit<scalar_t>::m) {
+
+    constexpr scalar_t inv{detail::invalid_value<scalar_t>()};
+    constexpr scalar_t epsilon{std::numeric_limits<scalar_t>::epsilon()};
+
+    // First, try to bracket a root
+    auto f = [&evaluate_func](const scalar_t x) {
+        auto [f_x, df_x] = evaluate_func(x);
+
+        return f_x;
+    };
+
+    // Initial bracket
+    scalar_t a{math::fabs(s) == 0.f ? -0.1f : 0.9f * s};
+    scalar_t b{math::fabs(s) == 0.f ? 0.1f : 1.1f * s};
+    std::array<scalar_t, 2> br{};
+    bool is_bracketed = expand_bracket(a, b, f, br);
+
+    // Update initial guess on the root after bracketing
+    s = is_bracketed ? 0.5f * (br[1] + br[0]) : s;
+
+    if (is_bracketed) {
+        // Check bracket
+        [[maybe_unused]] auto [f_a, df_a] = evaluate_func(br[0]);
+        [[maybe_unused]] auto [f_b, df_b] = evaluate_func(br[1]);
+
+        assert(math::signbit(f_a * f_b) && "Incorrect bracket around root");
+
+        // Root is not within the maximal pathlength
+        bool bracket_outside_tol{s > max_path &&
+                                 ((br[0] < -max_path && br[1] < -max_path) ||
+                                  (br[0] > max_path && br[1] > max_path))};
+        if (bracket_outside_tol) {
+#ifdef DEBUG
+            std::cout << "INFO: Root outside maximum search area - skipping"
+                      << std::endl;
+#endif
```

Comment:
Remember it's `#ifndef NDEBUG` rather than `#ifdef DEBUG`.

Reply:
Oops

Reply:
Done